File: common/java/src/main/java/org/aiddl/common/optimization/combinatorial/tsp/TspSolver.java
package org.aiddl.common.optimization.combinatorial.tsp;

import org.aiddl.common.CommonTerm;
import org.aiddl.common.math.graph.GraphTerm;
import org.aiddl.common.search.TreeSearch;
import org.aiddl.core.interfaces.Function;
import org.aiddl.core.representation.ListTerm;
import org.aiddl.core.representation.NumericalTerm;
import org.aiddl.core.representation.Term;
import org.aiddl.core.tools.LockableList;

public class TspSolver implements Function {

	@Override
	public Term apply(Term tsp) {
		TspMinRemainder remainder = new TspMinRemainder();
		remainder.initialize(tsp);
		
		TspExpandPath expand = new TspExpandPath();
		expand.initialize(tsp);
		
		TreeSearch search = new TreeSearch();
			
		Term path = Term.list();
		
		NumericalTerm best = Term.infPos();
		Term argBest = null;
		
		int numNodes = tsp.get(GraphTerm.Nodes).size();
		
		int a_count = 0;
		
		while ( !path.equals(CommonTerm.NIL) ) {
			Term expansion = expand.apply(path);
			a_count += 1;
			path = search.apply(Term.tuple(Term.sym("expand"), expansion));
			NumericalTerm pathCost = this.pathCost(path.asList());
			pathCost = remainder.apply(path).asNum().add(pathCost);
			
			while ( !path.equals(CommonTerm.NIL) && pathCost.greaterThan(best) ) {
				path = search.apply(Term.tuple(Term.sym("next")));
				if ( !path.equals(CommonTerm.NIL) ) {
					pathCost = this.pathCost(path.asList());
					pathCost = remainder.apply(path).asNum().add(pathCost);
				}
				a_count += 1;
			}
			if ( !path.equals(CommonTerm.NIL) && path.size() == numNodes && pathCost.lessThan(best) ) {
				best = pathCost;
				argBest = path;
			}
			
		}
		
		LockableList r = new LockableList();
		for ( Term e : argBest.asList() ) {
			r.add(e.get(1).getKey());
		}
		return Term.tuple(
				Term.keyVal(Term.sym("path"), Term.list(r)),
				Term.keyVal(Term.sym("cost"), best));

	}

	
	private NumericalTerm pathCost( ListTerm path ) {
		NumericalTerm sum = Term.integer(0);
		for ( Term e : path ) {
			sum = sum.add(e.get(0).asNum() );
		}
		return sum;
	}
}

File: common/scala/src/main/scala/org/aiddl/common/scala/optimization/combinatorial/tsp/TspSolver.scala
package org.aiddl.common.scala.optimization.combinatorial.tsp

import org.aiddl.common.scala.Common.NIL
import org.aiddl.common.scala.math.graph.GraphType.Undirected
import org.aiddl.common.scala.math.graph.{AdjacencyListGraph, Graph, Graph2Dot}
import org.aiddl.common.scala.math.graph.Terms.*
import org.aiddl.common.scala.search.GenericTreeSearch
import org.aiddl.core.scala.function.{Function, Initializable, InterfaceImplementation}
import org.aiddl.core.scala.representation.*

import scala.collection.immutable.HashSet
import scala.util.Random

import sys.process.*
import scala.language.postfixOps

class TspSolver extends GenericTreeSearch[Term, Term] with Initializable {
    val f_expand = new PathExpander()
    val f_minRemainder = new MinRemainder()
    var g: Graph = _
    var graphTerm: Term = _
    allowEarlyCostPruning = true

    override def init( g: Term ) = {
        this.graphTerm = g
        this.g = new AdjacencyListGraph(g)
        f_expand.init(this.g)
        f_minRemainder.init(this.g)
        super.reset
    }

    override def expand: Option[Seq[Term]] =
        f_expand(choice)

    override def cost( a: List[Term] ): Option[Num] =
        val cost =
            if a.length == 1
            then Num(0)
            else {
                val c = a.zip(a.tail).foldLeft(Num(0))((c, edge) => {
                    val (a, b) = edge
                    c + this.g.weight(a, b).get
                })
                val loopback =
                    if a.length != this.g.nodes.length
                    then Num(0)
                    else this.g.weight(a.reverse.head, a.head).get
                c + loopback
            }

        Some(f_minRemainder(a) + cost) // a.foldLeft(Num(0))( (c, v) => c + g.weight(v.asKvp.key, v.asKvp.value).get))

    override val nil: Term = ListTerm.empty

    override def assembleSolution(choice: List[Term]): Option[Term] =
        Some(ListTerm(choice))
}